Automattic/wp-calypso
# `Theme`: Use intersection observer to render a placeholder when not visible

Related to https://github.com/Automattic/wp-calypso/issues/95248

## Proposed Changes

* Update the `theme` component so it's lazy loaded when it's visible or within the next viewport height.

## Why are these changes being made?


* The `/themes` page loads ~34MB of theme screenshots on page load, because it's rendering 100 themes for each page. This is too much, and lazy loading the component allows us to reduce that to ~5MB.

* The initial [PR](https://github.com/Automattic/wp-calypso/pull/95881) had to be [reverted](https://github.com/Automattic/wp-calypso/pull/95951) because it was breaking an E2E test, which I fixed on https://github.com/Automattic/wp-calypso/pull/95984/commits/562d8b9afb871a3c535e0a469dc53850e8942c4a.

* I also noticed that the lazy loading was not working properly on the logged in theme showcase, with only the visible themes being loaded. This happened because the intersection observer's default root element is the document viewport, and since the logged in theme showcase has its own scroll, the intersection is not properly calculated. In order to fix it, I provided the theme showcase wrapper ref through a react context.

## Testing Instructions



* On the `Files changed` tab, enable the `Hide whitespace` option so the diff is more readable

* Open the live preview while logged out
* Open the dev tools, go to the network tab, filter by `.png`
* Go to `/themes`
* Check that only 8 theme screenshots are loaded (could be more, depending on the screen size), resulting in way less image data transferred (`5.1MB`) than what's happening on prod (`33.6MB`):
* Scroll down, check that the theme images are loaded 
* Check that it works as well for mobile devices
* Repeat the previous steps on the logged in theme showcase

|Before|After|
|---|---|
|![Untitled](https://github.com/user-attachments/assets/af9dcf72-ef82-43d0-ab57-27c64be6a161)|![Untitled](https://github.com/user-attachments/assets/0c5bfe75-3396-4c0a-979e-e949ee1194e0)|

## Pre-merge Checklist
